userdocs/diagnostics/associated-type-requirements.md

```diff
@@ -0,0 +1,18 @@
+# Using Protocols with Self or Associated Type Requirements
+
+A protocol with `Self` or associated type requirements may only be used to constrain the type of a generic parameter. This is because if a protocol named `Foo` has a requirement which references either `Self` or an associated type, then the type spelled `Foo` does not conform to the protocol `Foo` because it is unable to satisfy that requirement. As a result, `Foo` cannot be used as the type of a variable, argument, return value, etc. It may be used to constrain the type of a generic parameter `Bar` which represents a specific type.
```

Comment:
When you mentioned "cannot be used as the type of a ... return value" I was thinking that maybe opaque result types could be mentioned, not in a detailed way but maybe ... just mention that as we could have something like `func f() -> some Foo`?  Just a thought, maybe it makes sense :))

Reply:
IHMO my suggestion to talk about how referencing names in such contexts is a reference to a value of a particular type might be sufficient because opaque result type is a value to express exactly that in syntax...